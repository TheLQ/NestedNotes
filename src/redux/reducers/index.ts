import {
	AnyAction,
	// combineReducers,
	Reducer,
} from "redux";

import { AppDataModel } from "../../model/appData";
import { TagModel } from "../../model/tag";
import { UserDataModel } from "../../model/userData";

// import { InitAction, initUserDataReducer } from "./init";

export const INIT = "INIT";
// type TypeConsts = "INIT" | "ADD_ITEM";

export const ALL_REDUCERS: Reducer<AppDataModel> = function reducerWraper(
	state: AppDataModel,
	action: AnyAction): AppDataModel {
	switch (action.type) {
		case INIT:
			// return initUserDataReducer(state, action);
			return state;
		default:
			return state;
	}
};

export function initUserData(userData: UserDataModel): any {
	return {
		type: INIT,
		userData,
	};
}

export function activeTag(tag: TagModel): any {

	/*** TODO */
	// return {
	// 	type: INIT,
	// 	userData,
	// };
}
