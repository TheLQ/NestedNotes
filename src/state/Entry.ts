export interface Entry {
	readonly id: string;
}
