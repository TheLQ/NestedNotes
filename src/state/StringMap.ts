export interface StringMap<T> {
	[id: string]: T;
}

export interface ActiveStringMap<T> {
	entries: StringMap<T>;
	active: string | null;
}
