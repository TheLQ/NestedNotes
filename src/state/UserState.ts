import { ItemState } from "./ItemState";
import { TagState } from "./TagState";

export type NoteList = IdList<ItemState>;
export type TagList = IdList<TagState>;

export interface UserState {
	notes: NoteList;
	rootNotes: string[];
	tags: TagList;
}

export interface IdList<T> {
	[id: string]: T;
}
