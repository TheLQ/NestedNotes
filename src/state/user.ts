import { ItemState } from "./item";
import { TagState } from "./tag";

export type NoteList = IdList<ItemState>;
export type TagList = IdList<TagState>;

export interface UserState {
	notes: NoteList;
	rootNotes: string[];
	tags: TagList;
}

export interface IdList<T> {
	[id: string]: T;
}
