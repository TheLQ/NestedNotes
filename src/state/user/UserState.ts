import { StringMap } from "../StringMap";
import { BookState } from "./BookState";
import { UserViewState } from "./UserViewState";

export interface UserState {
	books: BookMap;
	name: string;
	views: UserViewMap;
}

export type BookMap = StringMap<BookState>;
export type UserViewMap = StringMap<UserViewState>;
