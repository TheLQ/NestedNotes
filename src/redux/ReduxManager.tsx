import { insertBelow, newEditor } from "./reducers/actions/EditorActions";
import React from "react";
import ReactDOM from "react-dom";
import { Provider } from "react-redux";
import { Action, applyMiddleware, createStore, Dispatch, Middleware, MiddlewareAPI, Store } from "redux";
import { composeWithDevTools } from "remote-redux-devtools";

import { RootState } from "../state/RootState";
import { UserState } from "../state/user/UserState";
import { validate } from "../state/Validator";
import { importUserData } from "../storage/StorageConvert";
import { ShellComponent } from "../ui/ShellComponent";
import { initUser } from "./reducers/actions/GeneralActions";
import {
	moveDown,
	moveLeft,
	moveRight,
	moveUp,
	selectNextActiveView,
	selectPrevActiveView,
} from "./reducers/actions/ViewActions";
import { RootReducer } from "./reducers/RootReducer";

export type RootStore = Store<RootState>;

//  extends {} is a workaround for limited generics in tsx
const crashReporter: Middleware = <S extends {}>(api: MiddlewareAPI<S>) =>
	(next: Dispatch<S>) =>
		<A extends Action>(action: A) => {
			try {
				return next(action);
			} catch (err) {
				// console.error('Caught an exception!', err)
				throw err;
			}
		};

const validator: Middleware = <S extends {}>(api: MiddlewareAPI<S>) =>
	(next: Dispatch<S>) =>
		<A extends Action>(action: A) => {
			const result = next(action);

			try {
				validate(api.getState() as {} as RootState);
			} catch (e) {
				console.error("failed post validate");
				throw e;
				// throw new Error(`failed post validate ${e}`);
			}

			return result;
		};

function defaultStoreFactory(): RootStore {
	// workaround for strict ts compiler
	const windowAny: any = window;

	const composeEnhancers = windowAny.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__
		? windowAny.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__({
			// Specify extensionâ€™s options like name, actionsBlacklist, actionsCreators, serialize...
		})
		: composeWithDevTools({});

	return createStore(
		RootReducer,
		composeEnhancers(
			applyMiddleware(crashReporter, validator),
		),
	) as RootStore;
}

export class ReduxManager {
	public readonly store: RootStore;

	public constructor(
		enableKeyboard: boolean,
		storeFactory: () => RootStore = defaultStoreFactory,
	) {
		this.store = storeFactory();
		if (enableKeyboard) {
			setSelectionKeyPressListener(this.store);
		}
	}

	private render() {
		return (
			<Provider store={ this.store } >
				<ShellComponent />
			</Provider>
			);
	}

	public create(element: HTMLElement) {
		if (element === null) {
			throw new Error("missing root element");
		}
		ReactDOM.render(this.render(), element);
	}

	public onUserDataLoad(fileState: UserState) {
		fileState = importUserData(fileState);
		this.store.dispatch(initUser(fileState));
	}
}

function selectionKeyPressListener(e: KeyboardEvent, store: RootStore) {
	console.log("keypress", e);

	// Do not trigger when inside a form
	const activeElement = document.activeElement as any;
	// console.log("activeElement", document.activeElement);
	if (activeElement.form !== undefined) {
		console.log("ignoring form");

		return;
	}

	switch (e.key) {
		// w
		case "w":
			store.dispatch(selectPrevActiveView());
			break;
		// shift + w
		case "W":
			store.dispatch(moveUp());
			break;
		// s
		case "s":
			store.dispatch(selectNextActiveView());
			break;
		// shift + s
		case "S":
			store.dispatch(moveDown());
			break;
		// shift + a
		case "A":
			store.dispatch(moveLeft());
			break;
		// shift + d
		case "D":
			store.dispatch(moveRight());
			break;
		// e
		case "e":
			// stop char from being inserted into newly created form element with focus
			e.preventDefault();
			store.dispatch(newEditor());
			break;
		case "ArrowDown":
			store.dispatch(insertBelow());
			break;
		default:
	}
}

let activeSelectionKeyPressListener: (e: KeyboardEvent) => void | undefined;
function setSelectionKeyPressListener(store: RootStore) {
	if (activeSelectionKeyPressListener !== undefined) {
		document.removeEventListener("keydown", activeSelectionKeyPressListener);
	}

	activeSelectionKeyPressListener = (e) => selectionKeyPressListener(e, store);
	document.addEventListener("keydown", activeSelectionKeyPressListener);
}
