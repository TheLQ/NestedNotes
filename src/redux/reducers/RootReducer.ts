import { AnyAction } from "redux";

import { initialState, RootState } from "../../state/RootState";

import { ClientReducer } from "./client/ClientReducer";
import { UserReducer } from "./user/UserReducer";

export function RootReducer(state: RootState = initialState, rawAction: AnyAction): RootState {
	const newUserState = UserReducer(state.user, rawAction);

	state = {
		user: newUserState,
		client: ClientReducer(state.client, newUserState, rawAction),
	};

	return state;
}
