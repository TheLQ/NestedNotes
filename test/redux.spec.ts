import jsonfile from "jsonfile";
import lodash from "lodash";
import { applyMiddleware, createStore } from "redux";
import { composeWithDevTools } from "remote-redux-devtools";

import {
	moveLeft,
	moveRight,
	selectNextActiveView,
	selectPrevActiveView,
} from "../src/redux/reducers/actions/ViewActions";
import { RootReducer } from "../src/redux/reducers/RootReducer";
import { ReduxManager, RootStore } from "../src/redux/ReduxManager";
import { Entry } from "../src/state/Entry";
import { ActiveStringMap } from "../src/state/StringMap";
import { UserState } from "../src/state/user/UserState";

class TestReduxManager extends ReduxManager {
	public constructor() {
		super(false, () => {
			const composeEnhancers = composeWithDevTools({
				realtime: true,
				hostname: "127.0.0.1",
				port: 8000, // the port your remotedev server is running at
			});

			return createStore(
				RootReducer,
				composeEnhancers(
					applyMiddleware(),
				),
			) as RootStore;
		});
	}
}

function reduxTest(test: (store: RootStore) => void): jest.ProvidesCallback {
	return (done) => {
		const div = document.createElement("div");
		const reduxManager = new ReduxManager(false);
		reduxManager.create(div);

		jsonfile.readFile("./dist/default.json", function(err: Error, data: UserState) {
			if (err) {
				throw err;
			}
			reduxManager.onUserDataLoad(data);
			try {
				test(reduxManager.store);
			} catch (e) {
				jsonfile.writeFile(
					"error-state.json",
					reduxManager.store.getState(),
					{spaces: 2},
					() => {
						done();
					},
				);
				// console.log("state");
				throw e;
			}
			done();
		});
	};
}

function active<T extends Entry>(map: ActiveStringMap<T>): T {
	if (map.active === undefined) {
		throw new Error("active is undefined");
	}

	const result = map.entries[map.active];
	if (result === undefined) {
		const keys = Object.keys(map.entries);
		const that = lodash.join(keys);
		throw new Error(`active ${map.active} does not exist in ${that} length ${keys.length}`);
	}

	return result;
}

function getActiveItem(store: RootStore) {
	const activeView = active(store.getState().client.views);
	const activeItem = active(activeView.items);

	return activeItem.id;
}

function expectStateDoesNotChange(store: RootStore, test: () => void) {
	const stateBefore = store.getState();
	test();
	expect(store.getState()).toEqual(stateBefore);
}

test("opens without crashing", reduxTest(() => {
	// do nothing
}));

test("select next/prev full", reduxTest((store) => {
	expect(getActiveItem(store)).toBe("item1");
	store.dispatch(selectNextActiveView());
	expect(getActiveItem(store)).toBe("item11");
	store.dispatch(selectNextActiveView());
	expect(getActiveItem(store)).toBe("item111");
	store.dispatch(selectNextActiveView());
	expect(getActiveItem(store)).toBe("item12");
	store.dispatch(selectNextActiveView());
	expect(getActiveItem(store)).toBe("item13");
	store.dispatch(selectNextActiveView());
	expect(getActiveItem(store)).toBe("item2");

	store.dispatch(selectPrevActiveView());
	expect(getActiveItem(store)).toBe("item13");
	store.dispatch(selectPrevActiveView());
	expect(getActiveItem(store)).toBe("item12");
	store.dispatch(selectPrevActiveView());
	expect(getActiveItem(store)).toBe("item111");
	store.dispatch(selectPrevActiveView());
	expect(getActiveItem(store)).toBe("item11");
	store.dispatch(selectPrevActiveView());
	expect(getActiveItem(store)).toBe("item1");
}));

describe("move inside the tree", () => {
	test("right", reduxTest((store) => {
		store.dispatch(selectNextActiveView());
		store.dispatch(selectNextActiveView());
		store.dispatch(selectNextActiveView());
		store.dispatch(selectNextActiveView());
		expect(getActiveItem(store)).toBe("item13");

		store.dispatch(moveRight());
		// expect(store.getState()).toMatchSnapshot();

		expectStateDoesNotChange(store, () => {
			store.dispatch(moveRight());
		});
	}));
	test("right at end does nothing", reduxTest((store) => {
		store.dispatch(selectNextActiveView());
		store.dispatch(selectNextActiveView());
		expect(getActiveItem(store)).toBe("item111");

		expectStateDoesNotChange(store, () => {
			store.dispatch(moveRight());
		});
	}));
	test("left", reduxTest((store) => {
		store.dispatch(selectNextActiveView());
		store.dispatch(selectNextActiveView());
		expect(getActiveItem(store)).toBe("item111");

		store.dispatch(moveLeft());
		// expect(store.getState()).toMatchSnapshot();
	}));
});

describe("move at root", () => {
	test.only("left to roots", reduxTest((store) => {
		store.dispatch(selectNextActiveView());
		expect(getActiveItem(store)).toBe("item11");

		store.dispatch(moveLeft());
		// expect(store.getState()).toMatchSnapshot();
	}));
	test.only("right to inside", reduxTest((store) => {
		store.dispatch(selectNextActiveView());
		store.dispatch(selectNextActiveView());
		store.dispatch(selectNextActiveView());
		store.dispatch(selectNextActiveView());
		store.dispatch(selectNextActiveView());
		expect(getActiveItem(store)).toBe("item2");

		store.dispatch(moveRight());
		// expect(store.getState()).toMatchSnapshot();
	}));
});
