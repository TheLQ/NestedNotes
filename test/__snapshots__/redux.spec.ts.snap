// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`move inside the tree left 1`] = `
Object {
  "client": Object {
    "views": Object {
      "active": "view1",
      "entries": Object {
        "view1": Object {
          "id": "view1",
          "items": Object {
            "active": "item111",
            "entries": Object {
              "item1": Object {
                "childNotes": Array [
                  "item11",
                  "item111",
                  "item12",
                  "item13",
                ],
                "id": "item1",
                "links": Array [],
                "settings": Object {},
                "tags": Array [
                  "top",
                ],
                "text": "test item 1",
              },
              "item11": Object {
                "childNotes": Array [],
                "id": "item11",
                "links": Array [],
                "parent": "item1",
                "settings": Object {},
                "tags": Array [
                  "child",
                ],
                "text": "test item 11",
              },
              "item111": Object {
                "childNotes": Array [],
                "id": "item111",
                "links": Array [],
                "parent": "item1",
                "settings": Object {},
                "tags": Array [
                  "grandchild",
                ],
                "text": "test item 111",
              },
              "item12": Object {
                "childNotes": Array [],
                "id": "item12",
                "links": Array [],
                "parent": "item1",
                "settings": Object {},
                "tags": Array [
                  "child",
                ],
                "text": "test item 12",
              },
              "item13": Object {
                "childNotes": Array [],
                "id": "item13",
                "links": Array [],
                "parent": "item1",
                "settings": Object {},
                "tags": Array [
                  "child",
                ],
                "text": "test item 13",
              },
              "item2": Object {
                "childNotes": Array [],
                "id": "item2",
                "links": Array [],
                "settings": Object {},
                "tags": Array [
                  "top",
                ],
                "text": "test item 2",
              },
            },
            "roots": Array [
              "item1",
              "item2",
            ],
          },
          "tags": Object {
            "active": undefined,
            "entries": Object {
              "child": Object {
                "id": "child",
                "name": "child",
              },
              "grandchild": Object {
                "id": "grandchild",
                "name": "grandchild",
              },
              "top": Object {
                "id": "top",
                "name": "top",
              },
            },
            "roots": Array [
              "top",
              "child",
              "child",
              "child",
              "grandchild",
              "top",
            ],
          },
        },
      },
    },
  },
  "user": Object {
    "books": Object {
      "firstbook": Object {
        "id": "firstbook",
        "items": Object {
          "entries": Object {
            "item1": Object {
              "childNotes": Array [
                "item11",
                "item12",
                "item13",
              ],
              "id": "item1",
              "links": Array [],
              "settings": Object {},
              "tags": Array [
                "top",
              ],
              "text": "test item 1",
            },
            "item11": Object {
              "childNotes": Array [
                "item111",
              ],
              "id": "item11",
              "links": Array [],
              "parent": "item1",
              "settings": Object {},
              "tags": Array [
                "child",
              ],
              "text": "test item 11",
            },
            "item111": Object {
              "childNotes": Array [],
              "id": "item111",
              "links": Array [],
              "parent": "item11",
              "settings": Object {},
              "tags": Array [
                "grandchild",
              ],
              "text": "test item 111",
            },
            "item12": Object {
              "childNotes": Array [],
              "id": "item12",
              "links": Array [],
              "parent": "item1",
              "settings": Object {},
              "tags": Array [
                "child",
              ],
              "text": "test item 12",
            },
            "item13": Object {
              "childNotes": Array [],
              "id": "item13",
              "links": Array [],
              "parent": "item1",
              "settings": Object {},
              "tags": Array [
                "child",
              ],
              "text": "test item 13",
            },
            "item2": Object {
              "childNotes": Array [],
              "id": "item2",
              "links": Array [],
              "settings": Object {},
              "tags": Array [
                "top",
              ],
              "text": "test item 2",
            },
          },
          "roots": Array [
            "item1",
            "item2",
          ],
        },
      },
    },
    "name": "test",
    "views": Object {
      "view1": Object {
        "forBookId": "firstbook",
        "id": "view1",
      },
    },
  },
}
`;

exports[`move inside the tree right 1`] = `
Object {
  "client": Object {
    "views": Object {
      "active": "view1",
      "entries": Object {
        "view1": Object {
          "id": "view1",
          "items": Object {
            "active": "item13",
            "entries": Object {
              "item1": Object {
                "childNotes": Array [
                  "item11",
                  "item12",
                ],
                "id": "item1",
                "links": Array [],
                "settings": Object {},
                "tags": Array [
                  "top",
                ],
                "text": "test item 1",
              },
              "item11": Object {
                "childNotes": Array [
                  "item111",
                ],
                "id": "item11",
                "links": Array [],
                "parent": "item1",
                "settings": Object {},
                "tags": Array [
                  "child",
                ],
                "text": "test item 11",
              },
              "item111": Object {
                "childNotes": Array [],
                "id": "item111",
                "links": Array [],
                "parent": "item11",
                "settings": Object {},
                "tags": Array [
                  "grandchild",
                ],
                "text": "test item 111",
              },
              "item12": Object {
                "childNotes": Array [
                  "item13",
                ],
                "id": "item12",
                "links": Array [],
                "parent": "item1",
                "settings": Object {},
                "tags": Array [
                  "child",
                ],
                "text": "test item 12",
              },
              "item13": Object {
                "childNotes": Array [],
                "id": "item13",
                "links": Array [],
                "parent": "item12",
                "settings": Object {},
                "tags": Array [
                  "child",
                ],
                "text": "test item 13",
              },
              "item2": Object {
                "childNotes": Array [],
                "id": "item2",
                "links": Array [],
                "settings": Object {},
                "tags": Array [
                  "top",
                ],
                "text": "test item 2",
              },
            },
            "roots": Array [
              "item1",
              "item2",
            ],
          },
          "tags": Object {
            "active": undefined,
            "entries": Object {
              "child": Object {
                "id": "child",
                "name": "child",
              },
              "grandchild": Object {
                "id": "grandchild",
                "name": "grandchild",
              },
              "top": Object {
                "id": "top",
                "name": "top",
              },
            },
            "roots": Array [
              "top",
              "child",
              "child",
              "child",
              "grandchild",
              "top",
            ],
          },
        },
      },
    },
  },
  "user": Object {
    "books": Object {
      "firstbook": Object {
        "id": "firstbook",
        "items": Object {
          "entries": Object {
            "item1": Object {
              "childNotes": Array [
                "item11",
                "item12",
                "item13",
              ],
              "id": "item1",
              "links": Array [],
              "settings": Object {},
              "tags": Array [
                "top",
              ],
              "text": "test item 1",
            },
            "item11": Object {
              "childNotes": Array [
                "item111",
              ],
              "id": "item11",
              "links": Array [],
              "parent": "item1",
              "settings": Object {},
              "tags": Array [
                "child",
              ],
              "text": "test item 11",
            },
            "item111": Object {
              "childNotes": Array [],
              "id": "item111",
              "links": Array [],
              "parent": "item11",
              "settings": Object {},
              "tags": Array [
                "grandchild",
              ],
              "text": "test item 111",
            },
            "item12": Object {
              "childNotes": Array [],
              "id": "item12",
              "links": Array [],
              "parent": "item1",
              "settings": Object {},
              "tags": Array [
                "child",
              ],
              "text": "test item 12",
            },
            "item13": Object {
              "childNotes": Array [],
              "id": "item13",
              "links": Array [],
              "parent": "item1",
              "settings": Object {},
              "tags": Array [
                "child",
              ],
              "text": "test item 13",
            },
            "item2": Object {
              "childNotes": Array [],
              "id": "item2",
              "links": Array [],
              "settings": Object {},
              "tags": Array [
                "top",
              ],
              "text": "test item 2",
            },
          },
          "roots": Array [
            "item1",
            "item2",
          ],
        },
      },
    },
    "name": "test",
    "views": Object {
      "view1": Object {
        "forBookId": "firstbook",
        "id": "view1",
      },
    },
  },
}
`;
